pkg/policy/distillery.go

```diff
@@ -36,17 +37,18 @@ func newPolicyCache(repo *Repository, idmgr identitymanager.IDManager) *policyCa
 	return cache
 }
 
-// lookupOrCreate adds the specified Identity to the policy cache, with a reference
-// from the specified Endpoint, then returns the threadsafe copy of the policy.
-func (cache *policyCache) lookupOrCreate(identity *identityPkg.Identity) *cachedSelectorPolicy {
+// lookup returns the selector policy for the given identity. This will return an error
+// if the identity is not in use by any endpoint on the node. That can happen
+// in cases where an endpoint changed its identity during policy regeneration, where the regeneration
+// is still referring to the old identity.
+func (cache *policyCache) lookup(identity *identityPkg.Identity) (*cachedSelectorPolicy, error) {
 	cache.Lock()
 	defer cache.Unlock()
 	cip, ok := cache.policies[identity.ID]
 	if !ok {
-		cip = newCachedSelectorPolicy(identity)
-		cache.policies[identity.ID] = cip
+		return nil, fmt.Errorf("SelectorPolicy not found in cache for ID %d", identity.ID)
```

Comment:
This is the old error message, but we can also add a comment saying the endpoint has changed its identity and therefore regeneration has failed, and will be retried.

Reply:
Maybe the error msg as is like this is fine and the caller can detect this error msg upper in the stack and react accordingly to the changed identity of the endpoint? That may we can give a better log msg or a better retry msg. Is that what you're thinking?

Reply:
I'm pretty pragmatic here. Its hard from this code path to know if this is expected during an identity change, or if its because of a leak like eg. this one being fixed; https://github.com/cilium/cilium/pull/42662.

This error message is the old one from before https://github.com/cilium/cilium/commit/36c9bba3a9424fbeb89c1b1127790e55df909764, since this is effectively just a revert of that commit.

I agree we can likely do better. I think to fully be sure, we can further down check if the endpoint actually changed identity, and then log as warn/info - but if the endpoint had the correct identity, we can log as an error.

Reply:
@odinuge That sounds reasonable to me, do you want to address this in this PR?

Reply:
I can look at it, but I think I'd prefer if we keep this PR rather simple. I can also update the description to something highlighting the _if_ here as well;

```suggestion
		return nil, fmt.Errorf("SelectorPolicy not found in cache for ID %d, meaning that either the endpoint changed its identity during regeneration or there is an issue in the identity cache", identity.ID)
```

Reply:
Sure, we can do this in a followup, no need to block this PR.